Optimize CPU version performance of the nonzero function.

VitalyFedyunin opened the pull request:
Optimized CPU version of the nonzero. Now 2x faster (in avg.) than numpy.

Can be further optimized for 1D tensors and boolean tensors.

VitalyFedyunin commented:
Related to #14848

gchanan commented on aten/src/TH/generic/THTensorEvenMoreMath.cpp:
```diff
@@ -138,20 +138,35 @@ void THTensor_(nonzero)(THLongTensor *subscript, THTensor *tensor)
+  int64_t sizes[dimensions];
+  int64_t idx[dimensions];
+  for (i = 0; i < dimensions; ++i) {
+    sizes[i] = THTensor_sizeLegacyNoScalars(tensor, i);
```
This won't break, but doesn't make a whole lot of sense.  THTensor_sizeLegacyNoScalars is for getting 1 instead of 0 if you have a 0-dimensional tensor, but you can only enter this loop if it's not a 0-dimensional tensor.

VitalyFedyunin replied:
Resolved by adding return statement.

VitalyFedyunin commented:
![screen shot 2018-12-14 at 1 01 35 pm](https://user-images.githubusercontent.com/1163042/50019883-41e34b80-ffa2-11e8-8ee7-9a01baf06ba7.png)

New version with pointers math works faster than numpy in all cases (2x faster in avg.).

Still need to change GPU version.

VitalyFedyunin commented on aten/src/TH/generic/THTensorEvenMoreMath.cpp:
```diff
@@ -138,20 +138,35 @@ void THTensor_(nonzero)(THLongTensor *subscript, THTensor *tensor)
   THLongTensor_resize2d(subscript, numel, tensor->dim());
 
+  int64_t dimensions = tensor->dim();
+  int64_t sizes[dimensions];
```
Now need it in reverse order.

gchanan commented on aten/src/TH/generic/THTensorEvenMoreMath.cpp:
```diff
@@ -138,20 +136,43 @@ void THTensor_(nonzero)(THLongTensor *subscript, THTensor *tensor)
+                    ii = idx + dimensions;
+                    for (int64_t dim = dimensions - 1; dim >= 0; dim--) {
+                      --ii;
+                      *subscript_data = *ii;
```
this looks broken for the case where subscript is not resized (passed in as the correct size) and not contiguous, although it looks like it was broken before too.

For example:
```
>>> x
tensor([[1, 0, 1],
             [0, 1, 1]], dtype=torch.uint8)

>>> torch.nonzero(x)
tensor([[0, 0],
            [0, 2],
            [1, 1],
            [1, 2]])

>>> y=torch.full((4,2,2),5, dtype=torch.long)[:,:,1]
>>> torch.nonzero(x,out=y)
tensor([[0, 0],
             [1, 1],
             [5, 5],
             [5, 5]])
```

VitalyFedyunin commented:
@gchanan bug fixed

gchanan reviewed (approved):
A test for the non-contiguous case would be a nice addition.

This looks good to go though, feel free to either commit or address the comments then commit.

VitalyFedyunin committed:
Optimize CPU version performance of the nonzero function. (#15190)

Summary:
Optimized CPU version of the nonzero. Now 2x faster (in avg.) than numpy.

Can be further optimized for 1D tensors and boolean tensors.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15190

Differential Revision: D13468570

fbshipit-source-id: 31e155c5ef247a8983b4c1c12f25b0aafb315e43

VitalyFedyunin commented on aten/src/TH/generic/THTensorEvenMoreMath.cpp:
```diff
@@ -137,21 +135,43 @@ void THTensor_(nonzero)(THLongTensor *subscript, THTensor *tensor)
+  sizes[dimensions] = 0;
   /* Second pass populates subscripts */
   subscript_data = THLongTensor_data(subscript);
+  auto subscript_strides = THTensor_stridesLegacyNoScalars(subscript);
```
subscript_strides = subscript->strides().vec(); looks dangerous for me (relay on internal implementation).

gchanan replied:
that's a valid concern; TH is a little weird in that it operates on TensorImpl (the pImpl) so mapping things to the main API you would usually re-wrap the pImpl, i.e.:
```
Tensor subscript_(subscript);
auto strides = subscript_.strides().vec();
```
or use APIs at the `THTensor` level.

but this just wastes a incref / decref.  Perhaps a better alternative is just to create a `THTensor_strides(...)` function; there's no particular reason we have the noScalars version and not the main one.

botcs commented:
Hi @gchanan 
This still seems to be a problem in 1.1.0 for some applications.
If you want to have only the sum of nonzero elements it is faster by an order of magnitude to clamp your tensor and do a sum over it. Both on GPU (I used `torch.cuda.synchronize` for benchmarking) and CPU.

also using `uint8` as dtype instead of `int64` will double the computing time, that is really annoying

VitalyFedyunin commented:
@botcs are you talking about CPU or GPU implementations

gchanan commented:
@botcs please file an issue.

botcs commented:
I have to correct myself, the GPU time does not double, I had other processes stuck in. Sorry...

@gchanan filed an issue #23907

makslevental commented:
@VitalyFedyunin how can this be optimized for Boolean tensors?

VitalyFedyunin commented:
To be honest first step here should be migrating from TH to Aten code, as it might help with vectorization.
